simulation/src/io/WriterHDF5.cpp

```
@@ -0,0 +1,438 @@
+    , hdf5_Real_type {std::is_same<double, Real>::value ? H5T_NATIVE_DOUBLE : H5T_NATIVE_FLOAT}
+    , m_variables {variables}
+{
+    int tmp_rank=0;
+#if defined(MPI_SESSION)
+    MPI_Comm_rank(MPI_COMM_WORLD, &tmp_rank);
+    m_mpi_coords = grid.comm.getCoords(grid.comm.rank());
+#endif
+
+    std::ostringstream mpi_prefix;
+    mpi_prefix << std::setw(3) << std::setfill('0') << tmp_rank;    
+    m_prefix.append("_r"+mpi_prefix.str());
+}
+
+std::string WriterHDF5::getFilename(Int outputId) const{
+
+    // write outputId in string outputNum
+    std::ostringstream outputNum;
+    outputNum << std::setw(std::numeric_limits<Int>::digits10);
+    outputNum << std::setfill('0');
+    outputNum << outputId;
+
+    // concatenate file prefix + file number + suffix
+    std::string filename(m_prefix);
+    filename += "_" + outputNum.str();
+    filename += ".h5";
+    return filename;
+}
+
+// write simple dataset(scalar, vector, array)
+void WriterHDF5::write_simple_dataset_select(const hid_t &file_id, const char* name, const hid_t &type,
+    const hsize_t dims_size, const hsize_t dims[/*size*/],
+    const hsize_t hyperslab_start[/*size*/], const hsize_t hyperslab_count[/*size*/],
+    const void* data)
+{
+    // If we write all element in data,  hyperslab_start = 0  and hyperslab_count = dims
+    herr_t status=-1;
+    herr_t status_tmp=0;
+    hid_t dataset_id;
+    hid_t dataspace_id;
+    hid_t datamemory_id;
+
+    if (dims_size == 1 && dims[0] == 1){
+        dataspace_id = H5Screate(H5S_SCALAR);
+    }
+    else{
+        /*
+        * Describe the size of the array and create the data space for fixed
+        * size dataset.
+        */
+        dataspace_id = H5Screate_simple(dims_size, hyperslab_count, NULL);
+    }
+    if (dataspace_id < 0) {
+        std::cout << "error "<< dataspace_id <<" in creating dataspace for " << name << std::endl;
+        goto end_function;
```

Why use goto? 
Better to throw a runtime error if hdf5 fails at any time.